### DISCUSSION TITLE
Enable lowest OPUS framesize not only for <64 samples buffer.

### DISCUSSION THREAD

**Discussion by 38github** [jamulussoftware/jamulus#3040] [Technical Support & Help]
Is there a reason why the lowest OPUS framesize are not enabled when using other than <64 samples of latency buffer? I believe that being able to achieve the lowest delay even if you do not or can't use 64 samples is important to make the Jamulus experience as good as it can get.

These are the numbers I get when setting Jamulus' local
and server buffers to 1 and 1, enabling "Enable Small
Network Buffers" and excluding the ping time from the
delay value.

```
 64/2 =  8 ms delay (JACK reports  64/2 equals 1.3 ms)
128/2 = 16 ms delay (JACK reports 128/2 equals 2.7 ms)
256/2 = 26 ms delay (JACK reports 256/2 equals 5.3 ms)
```

This means that on top of the audio interface's
latencies:

```
 64/2 has  6.7 ms
128/2 has 13.3 ms
256/2 has 20.7 ms
```

Imagine if we instead could have these Jamulus delay
times:

```
 64/2 =  8 ms delay
128/2 =  9 ms delay
256/2 = 11 ms delay
```

Some of these numbers are approximates and/or rounded
up or down to a whole number.

If I am wrong, which I potentially might be, then
please correct me.

**Comment by pljones** [comment]
There are four sets of buffers:
- audio buffers
- network buffers ("Small network buffers" checkbox under the jitter buffers)
- the pair of jitter buffers

I'll skip the jitter buffers - they're in multiples of the network buffer size, as far as I'm aware but beyond that... I get a bit lost.

It's down to the client to pick the audio buffer size most appropriate to their system capabilities.

The server has to coordinate clients using network buffers of either 64 or 128 samples.  Basically, it takes two 64 sample buffers from a client using that size buffer to queue to all clients using 128 sample buffers and sends 64 sample buffers as soon as it can to clients using them.

Say network buffers of 32 samples were introduced.  The server would collect four such buffers from a client using that buffer size before queuing for mixing with other clients.  The mix for that client would then be a 128 sample block sent as four 32 sample buffers.  So there wouldn't be any time saving within the server.

So whilst the mix from the server would start to arrive sooner, it wouldn't have left any sooner.  (But with more codec and network overhead.)

The main point is that the server mixes in 128 sample blocks - nothing related to OPUS, just to keeping all clients in line and vaguely synchronised.

**Reply by 38github** [reply]
If I do 64/4 (instead of 128/2) and 64/8 (instead of 256/2) I get the benefits of lower delay. How come this works but using 128/2 and 256/2 creates larger delay?

**Reply by dcorson-ticino-com** [reply]
Please explain to me what the nomenclature "64/4" or "128/2" means.
Is that 4 buffers of 64 samples and 2 buffers of 128 samples or does it mean 16 samples and 64 samples, and if so why not just use the number?
To my knowledge Jamulus uses buffers of 64 samples and 128 samples.
Everything else is grouping of those basic buffers to lower the overhead and it never uses fractional buffers.

**Reply by 38github** [reply]
That is correct. Those are the JACK settings I used. Instead of using 128/2 I can trick Jamulus to think I am still using  64 by using 4/64 and I get the benefits of lower delay.

**Reply by dcorson-ticino-com** [reply]
I doubt that there is any real benefit of "tricking" Jamulus to think it is using 64 sample buffers.
Your operating system is still using a buffer of 256, otherwise there would be no reason to use the 128/2 or 64/4 nomenclature.
In Jamulus the ping times are really measured, but the delay times are calculated, not measured.
So if you tell Jamulus to use a 64 sample buffer it will calculate a better time even if in reality your operating system is using a 256 sample buffer and there is no improvement in reality.
Several years ago someone set up a test arrangement to really measure the delay times.  The results may still be available on SourceForge.

**Reply by 38github** [reply]
When using 64/4 instead of 128/2 Jamulus is activating "Enable Small
Network Buffers".

**Reply by dcorson-ticino-com** [reply]
> When using 64/4 instead of 128/2 Jamulus is activating "Enable Small Network Buffers".

But if your operating system is still sending 256 sample buffers that makes no difference.
I see no reason to use a nomenclature such as 64/4 other than to tell you that the system is still using 256, otherwise you would simply say the buffer is 64.

**Reply by 38github** [reply]
Please read and analyse/test what I write. Here is the proof:

![jamulus_difference](https://user-images.githubusercontent.com/18294149/229251043-b4d363c3-6fcb-4442-b80b-f1a430885e7f.png)

In Qjackctl I connected my mic to Reaper IN 1 and Jamulus IN 1. Then I connected Jamulus OUT 1 to Reaper IN 2. In Reaper I then recorded both the direct mic signal and the output of Jamulus to seperate tracks. They share the AD conversion latency but Jamulus also has the send and receive latency. I used Local 1 and Server 1 buffers in Jamulus to get the lowest delay time.

Here it shows that when Jamulus detects 64 samples of buffer it lowers the delay time (Enable Small Network Buffers). Even if the total amount equals 128 samples or more. My question is: Why not enable _Enable Small Network Buffers_ all the time if the server uses _--fastupdate_?

**Reply by pljones** [reply]
`--fastupdate` is required for _Small Network Buffers_ to have any effect anyway.  But it only affects clients with _Small Network Buffers_ enabled -- the server has to cater for clients without.  And the reason for that is the overheads are increased if you're creating more packets: some systems may work better **without** _Small Network Buffers_.

**Reply by 38github** [reply]
Can't the server, if configured with --fastupdate, have it active for 64, 128 and 256 clients? If the server is slow or the internet connection is slow they just have to not configure it with --fastupdate? Even old $20 SoCs can handle up to ten clients using 64 samples and Enable Small Network Buffers.

**Reply by pljones** [reply]
Receiving and sending network packets is what `--fastupdate` and `Small Network Buffers` control.  (And possibly the jitter buffer granularity.)

If you have 64 sample audio buffers and choose not to use small network buffers, you'll still see lower latency than when using larger audio buffers - but the latency will drop if you use small network buffers **and** you connect to a server with a high enough version configured with `--fastupdate`.

If you have audio buffers above 128 samples, small network buffers and `--fastupdate` won't have any effect.  There's nothing the server can do about that.

We don't have any information about which clients are using the latest software but we _know_ not all servers are.

So new clients need to support both old and new servers.

And new servers need to support a _combination_ of old clients and new clients at the same time.  To do that, they have to use a single sized buffer for the mix: that's the same size that's always been used - 128 samples.

To get the 128 samples with smaller network buffers, that means waiting for multiple incoming packets and queuing split buffers into smaller packets.

For "standard" network buffers, the mix buffer is loaded and queued directly from/to the network layer.

**Reply by dcorson-ticino-com** [reply]
> Can't the server, if configured with --fastupdate, have it active for 64, 128 and 256 clients? If the server is slow or the internet connection is slow they just have to not configure it with --fastupdate? Even old $20 SoCs can handle up to ten clients using 64 samples and Enable Small Network Buffers.

That depends more on the Internet provider than your hardware.  Two people I play with regularly are using latest Apple hardware and have much poorer sound (on the same -f supporting server) using small buffers that without.  Figure it out.

**Reply by 38github** [reply]
Then they have the option to disable -f. If I remember correctly it is not enabled by default either on client or server.

**Comment by jujudusud** [comment]
Wouldn't the main reason be that OPUS is only designed to use 128 samples buffers by default?

**Reply by pljones** [reply]
Getting into the codec itself is a different matter - that raises the question over whether a different codecs could be used where network latency allowed it.  The resultant audio could be mixed anyway, with the existing codec remaining available.

**Comment by 38github** [comment]
So, from the developers perspective, is this something that is possible to do?

**Reply by pljones** [reply]
It would need someone wanting to write the code.  Do you want to?

**Reply by 38github** [reply]
I am more interested at this point if it would be possible.

**Reply by pljones** [reply]
That would most depend on someone wanting to write the code.  Without that, then no, it's never going to happen.
